pkg/tests/tasks/traffic/traffic_tcp_shifting_test.go

```diff
@@ -15,100 +15,201 @@
+		app.InstallAndWaitReady(t, app.Sleep(ns), app.EchoV1(ns), app.EchoV2(ns))
+
+		t.NewSubTest("tcp shift 100 percent to v1").Run(func(t test.TestHelper) {
+			t.Cleanup(func() {
+				oc.DeleteFromString(t, ns, EchoAllv1Yaml)
+			})
+
+			t.LogStep("Shifting all TCP traffic to v1")
+			oc.ApplyString(t, ns, EchoAllv1Yaml)
+
+			t.LogStep("make 20 requests and checking if all of them go to v1 (tolerance: 0%)")
+			retry.UntilSuccess(t, func(t test.TestHelper) {
+				tolerance := 0.0
+				checkTcpTrafficRatio(t, ns, "tcp-echo", "9000", 20, tolerance, map[string]float64{
+					"one": 1.0,
+					"two": 0.0,
+				})
+			})
+		})
+
+		t.NewSubTest("tcp shift 20 percent to v2").Run(func(t test.TestHelper) {
+			t.Cleanup(func() {
+				oc.DeleteFromString(t, ns, Echo20v2Yaml)
+			})
+
+			t.LogStep("Shifting 20 percent TCP traffic to v2")
+			oc.ApplyString(t, ns, Echo20v2Yaml)
+
+			t.LogStep("make 100 requests and checking if 20 percent of them go to v2 (tolerance: 10%)")
+			retry.UntilSuccess(t, func(t test.TestHelper) {
+				tolerance := 0.10
+				checkTcpTrafficRatio(t, ns, "tcp-echo", "9000", 100, tolerance, map[string]float64{
+					"one": 0.8,
+					"two": 0.2,
+				})
+			})
+		})
 	})
+}
 
-	t.Run("TrafficManagement_20_percent_v2_tcp_shift_test", func(t *testing.T) {
-		defer util.RecoverPanic(t)
-
-		log.Log.Info("Shifting 20%% TCP traffic to v2 tolerance 15%%")
-		util.KubeApplyContents("bookinfo", echo20v2Yaml)
-
-		tolerance := 0.15
-		totalShot := 60
-		c1, c2 := 0, 0
-
-		sleepPod, err := util.GetPodName("bookinfo", "app=sleep")
-		util.Inspect(err, "Failed to get sleep pod name", "", t)
-		cmd := fmt.Sprintf(`sh -c "(date; sleep 1) | nc %s %s"`, "tcp-echo", "9000")
-
-		for i := 0; i < totalShot; i++ {
-			msg, err := util.PodExec("bookinfo", sleepPod, "sleep", cmd, true)
-			util.Inspect(err, "Failed to get response", "", t)
-			if strings.Contains(msg, "one") {
-				c1++
-			} else if strings.Contains(msg, "two") {
-				c2++
-			} else {
-				log.Log.Errorf("Unexpected echo version: %s", msg)
-			}
-		}
-		if util.IsWithinPercentage(c1, totalShot, 0.8, tolerance) && util.IsWithinPercentage(c2, totalShot, 0.2, tolerance) {
-			log.Log.Infof("Success. Traffic shifting acts as expected. "+
-				"v1 version hit %d, v2 version hit %d", c1, c2)
+func checkTcpTrafficRatio(t test.TestHelper, ns, host, port string, numberOfRequests int, tolerance float64, ratios map[string]float64) {
+	counts := map[string]int{}
+	for i := 0; i < numberOfRequests; i++ {
+		oc.Exec(t,
+			pod.MatchingSelector("app=sleep", ns),
+			"sleep",
+			fmt.Sprintf(`sh -c "(date; read -t 0.5) | nc %s %s"`, host, port),
+			func(t test.TestHelper, output string) {
+				matched := false
+				for version := range ratios {
+					if strings.Contains(output, version) {
+						matched = true
+						counts[version]++
+					}
+				}
+				if !matched {
+					t.Fatalf("nc tcp-echo output did not match any expected version, got output: %s", output)
+				}
+			})
+	}
```

Comment:
The reason this is slow is because `oc.Exec` itself takes a long time, regardless of the command. So, why not do something like this:
```suggestion
	oc.Exec(t,
		pod.MatchingSelector("app=sleep", ns),
		"sleep",
		fmt.Sprintf(`sh -c 'i=1; while [ $i -le %d ]; do date | nc %s %s; i=$((i+1)); done'`, numberOfRequests, host, port),
		func(t test.TestHelper, output string) {
			scanner := bufio.NewScanner(strings.NewReader(output))
			for scanner.Scan() {
				line := scanner.Text()
				matched := false
				for version := range ratios {
					if strings.Contains(line, version) {
						matched = true
						counts[version]++
					}
				}
				if !matched {
					t.Fatalf("nc tcp-echo output did not match any expected version, got output: %s", output)
				}
			}
		})
```

This way, 100 requests are made in less than a second, and the entire test takes around 15s.